### ISSUE TITLE
`suspend-then-hibernate` hangs when used with nvidia closed source, and behaves differently than `suspend`

### ISSUE THREAD

**Issue by soraxas** [systemd/systemd#27559]
### systemd version the issue has been seen with

253

### Used distribution

Manjaro rolling release

### Linux kernel version used

6.1.25-1-MANJARO

### CPU architectures issue was seen on

None

### Expected behaviour you didn't see

- `suspend-then-hibernate` can (at least) successfully suspend the system
- `suspend-then-hibernate` behaves similar to `suspend` + `hibernate`

### Unexpected behaviour you saw

`systemctl suspend` works, `systemctl hibernate` works, but `systemctl suspend-then-hibernate` hangs.

After running the command, the system hangs/freeze after the screen went black (with fan stills on). Pressing keyboard will not wake the system. I can use sysrq sequence to kill all tasks, of which the system will then spawn the DE again, and regain control of the computer.

### Steps to reproduce the problem

Running `sudo systemctl suspend-then-hibernate` or `sudo /usr/lib/systemd/systemd-sleep suspend-then-hibernate` both hangs the system.

After some investigation, I realize that the main difference between pure `suspend` and `suspend-then-hibernate` is that `suspend-then-hibernate` contains some extra function that freeze/thaw `user.slice` (introduced in #24336). (question, should this be specific to `suspend-then-hibernate`? or should this be applied to `suspend`/`hibernate` as well?)

https://github.com/systemd/systemd/blob/b3d12ac0da2a8b25abc9052a4f350f1ea1108985/src/sleep/sleep.c#L377-L397

During the hangs in `suspend-then-hibernate`, I've used sysrq to dump the blocked state with the following output (I assume it's during freezing user.slice)
```
May 07 19:43:41 xps9500 kernel: sysrq: Show Blocked State
May 07 19:43:41 xps9500 kernel: task:kworker/10:1    state:D stack:0     pid:133   ppid:2      flags:0x00004000
May 07 19:43:41 xps9500 kernel: Workqueue: pm pm_runtime_work
May 07 19:43:41 xps9500 kernel: Call Trace:
May 07 19:43:41 xps9500 kernel:  <TASK>
May 07 19:43:41 xps9500 kernel:  __schedule+0x370/0x12a0
May 07 19:43:41 xps9500 kernel:  ? __mod_timer+0x289/0x3b0
May 07 19:43:41 xps9500 kernel:  schedule+0x5e/0xd0
May 07 19:43:41 xps9500 kernel:  schedule_timeout+0x98/0x150
May 07 19:43:41 xps9500 kernel:  ? __bpf_trace_tick_stop+0x10/0x10
May 07 19:43:41 xps9500 kernel:  msleep+0x2d/0x40
May 07 19:43:41 xps9500 kernel:  acpi_ex_system_do_sleep+0x2b/0x40
May 07 19:43:41 xps9500 kernel:  acpi_ex_opcode_1A_0T_0R+0xe7/0x130
May 07 19:43:41 xps9500 kernel:  acpi_ds_exec_end_op+0x31e/0x850
May 07 19:43:41 xps9500 kernel:  acpi_ps_parse_loop+0x27b/0xa20
May 07 19:43:41 xps9500 kernel:  acpi_ps_parse_aml+0x221/0x5e0
May 07 19:43:41 xps9500 kernel:  acpi_ps_execute_method+0x171/0x3e0
May 07 19:43:41 xps9500 kernel:  acpi_ns_evaluate+0x2dc/0x5d0
May 07 19:43:41 xps9500 kernel:  acpi_evaluate_object+0x16f/0x450
May 07 19:43:41 xps9500 kernel:  __acpi_power_off+0x21/0x60
May 07 19:43:41 xps9500 kernel:  acpi_power_off_unlocked+0x48/0x80
May 07 19:43:41 xps9500 kernel:  acpi_power_off_list.isra.0+0x3d/0x90
May 07 19:43:41 xps9500 kernel:  acpi_power_transition+0x84/0xb0
May 07 19:43:41 xps9500 kernel:  acpi_device_set_power+0x2eb/0x410
May 07 19:43:41 xps9500 kernel:  acpi_pci_set_power_state+0x88/0x100
May 07 19:43:41 xps9500 kernel:  pci_set_power_state+0x9e/0x210
May 07 19:43:41 xps9500 kernel:  ? pcie_portdrv_mmio_enabled+0x20/0x20
May 07 19:43:41 xps9500 kernel:  pci_finish_runtime_suspend+0x45/0x80
May 07 19:43:41 xps9500 kernel:  ? pcie_portdrv_mmio_enabled+0x20/0x20
May 07 19:43:41 xps9500 kernel:  pci_pm_runtime_suspend+0xc9/0x1b0
May 07 19:43:41 xps9500 kernel:  ? __queue_work+0x1dd/0x430
May 07 19:43:41 xps9500 kernel:  ? pci_dev_put+0x20/0x20
May 07 19:43:41 xps9500 kernel:  __rpm_callback+0x41/0x170
May 07 19:43:41 xps9500 kernel:  ? pci_dev_put+0x20/0x20
May 07 19:43:41 xps9500 kernel:  rpm_callback+0x5d/0x70
May 07 19:43:41 xps9500 kernel:  ? pci_dev_put+0x20/0x20
May 07 19:43:41 xps9500 kernel:  rpm_suspend+0x11a/0x720
May 07 19:43:41 xps9500 kernel:  ? pci_reassigndev_resource_alignment+0x2b0/0x2b0
May 07 19:43:41 xps9500 kernel:  pm_runtime_work+0x84/0xb0
May 07 19:43:41 xps9500 kernel:  process_one_work+0x1c4/0x380
May 07 19:43:41 xps9500 kernel:  worker_thread+0x51/0x390
May 07 19:43:41 xps9500 kernel:  ? rescuer_thread+0x3b0/0x3b0
May 07 19:43:41 xps9500 kernel:  kthread+0xdb/0x110
May 07 19:43:41 xps9500 kernel:  ? kthread_complete_and_exit+0x20/0x20
May 07 19:43:41 xps9500 kernel:  ret_from_fork+0x1f/0x30
May 07 19:43:41 xps9500 kernel:  </TASK>
```

I've actually tried to compile systemd from source by removing the freeze function, i.e.,
```c
/* Freeze when invoked and thaw on cleanup */
static int freeze_thaw_user_slice(const char **method) {
		return 0;
        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
        _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
        int r;
...
...
}
```
and then runs it via `sudo ./build/systemd-sleep suspend-then-hibernate`, which in my limited case seems to works ok (behaves similar to `suspend`, as expected)

see https://0x0.st/HZyb.log for full log

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
May 07 19:43:30 xps9500 dbus-daemon[760]: [system] Activating via systemd: service name='org.freedesktop.nm_dispatcher' unit='dbus-org.freedesktop.nm-dispatcher.service' requested by ':1.6' (uid=0 pid=819 comm="/usr/bin/NetworkManager --no-daemon")
May 07 19:43:30 xps9500 systemd[1]: Starting Network Manager Script Dispatcher Service...
May 07 19:43:30 xps9500 dbus-daemon[760]: [system] Successfully activated service 'org.freedesktop.nm_dispatcher'
May 07 19:43:30 xps9500 systemd[1]: Started Network Manager Script Dispatcher Service.
May 07 19:43:31 xps9500 kernel: ucsi_acpi USBC000:00: failed to re-enable notifications (-110)
May 07 19:43:32 xps9500 dbus-daemon[760]: [system] Activating via systemd: service name='org.freedesktop.home1' unit='dbus-org.freedesktop.home1.service' requested by ':1.149' (uid=0 pid=8664 comm="sudo /usr/lib/systemd/systemd-sleep suspend-then-hibernate")
May 07 19:43:32 xps9500 dbus-daemon[760]: [system] Activation via systemd failed for unit 'dbus-org.freedesktop.home1.service': Unit dbus-org.freedesktop.home1.service not found.
May 07 19:43:32 xps9500 sudo[8664]: pam_systemd_home(sudo:account): systemd-homed is not available: Unit dbus-org.freedesktop.home1.service not found.
May 07 19:43:32 xps9500 sudo[8664]:  soraxas : TTY=pts/2 ; PWD=/home/soraxas ; USER=root ; COMMAND=/usr/lib/systemd/systemd-sleep suspend-then-hibernate
May 07 19:43:32 xps9500 sudo[8664]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=1000)
May 07 19:43:40 xps9500 systemd[1]: NetworkManager-dispatcher.service: Deactivated successfully.
#############################
Alt-Sysrq-w
#############################
May 07 19:43:41 xps9500 kernel: sysrq: Show Blocked State
May 07 19:43:41 xps9500 kernel: task:kworker/10:1    state:D stack:0     pid:133   ppid:2      flags:0x00004000
May 07 19:43:41 xps9500 kernel: Workqueue: pm pm_runtime_work
May 07 19:43:41 xps9500 kernel: Call Trace:
May 07 19:43:41 xps9500 kernel:  <TASK>
May 07 19:43:41 xps9500 kernel:  __schedule+0x370/0x12a0
May 07 19:43:41 xps9500 kernel:  ? __mod_timer+0x289/0x3b0
May 07 19:43:41 xps9500 kernel:  schedule+0x5e/0xd0
May 07 19:43:41 xps9500 kernel:  schedule_timeout+0x98/0x150
May 07 19:43:41 xps9500 kernel:  ? __bpf_trace_tick_stop+0x10/0x10
May 07 19:43:41 xps9500 kernel:  msleep+0x2d/0x40
May 07 19:43:41 xps9500 kernel:  acpi_ex_system_do_sleep+0x2b/0x40
May 07 19:43:41 xps9500 kernel:  acpi_ex_opcode_1A_0T_0R+0xe7/0x130
May 07 19:43:41 xps9500 kernel:  acpi_ds_exec_end_op+0x31e/0x850
May 07 19:43:41 xps9500 kernel:  acpi_ps_parse_loop+0x27b/0xa20
May 07 19:43:41 xps9500 kernel:  acpi_ps_parse_aml+0x221/0x5e0
May 07 19:43:41 xps9500 kernel:  acpi_ps_execute_method+0x171/0x3e0
May 07 19:43:41 xps9500 kernel:  acpi_ns_evaluate+0x2dc/0x5d0
May 07 19:43:41 xps9500 kernel:  acpi_evaluate_object+0x16f/0x450
May 07 19:43:41 xps9500 kernel:  __acpi_power_off+0x21/0x60
May 07 19:43:41 xps9500 kernel:  acpi_power_off_unlocked+0x48/0x80
May 07 19:43:41 xps9500 kernel:  acpi_power_off_list.isra.0+0x3d/0x90
May 07 19:43:41 xps9500 kernel:  acpi_power_transition+0x84/0xb0
May 07 19:43:41 xps9500 kernel:  acpi_device_set_power+0x2eb/0x410
May 07 19:43:41 xps9500 kernel:  acpi_pci_set_power_state+0x88/0x100
May 07 19:43:41 xps9500 kernel:  pci_set_power_state+0x9e/0x210
May 07 19:43:41 xps9500 kernel:  ? pcie_portdrv_mmio_enabled+0x20/0x20
May 07 19:43:41 xps9500 kernel:  pci_finish_runtime_suspend+0x45/0x80
May 07 19:43:41 xps9500 kernel:  ? pcie_portdrv_mmio_enabled+0x20/0x20
May 07 19:43:41 xps9500 kernel:  pci_pm_runtime_suspend+0xc9/0x1b0
May 07 19:43:41 xps9500 kernel:  ? __queue_work+0x1dd/0x430
May 07 19:43:41 xps9500 kernel:  ? pci_dev_put+0x20/0x20
May 07 19:43:41 xps9500 kernel:  __rpm_callback+0x41/0x170
May 07 19:43:41 xps9500 kernel:  ? pci_dev_put+0x20/0x20
May 07 19:43:41 xps9500 kernel:  rpm_callback+0x5d/0x70
May 07 19:43:41 xps9500 kernel:  ? pci_dev_put+0x20/0x20
May 07 19:43:41 xps9500 kernel:  rpm_suspend+0x11a/0x720
May 07 19:43:41 xps9500 kernel:  ? pci_reassigndev_resource_alignment+0x2b0/0x2b0
May 07 19:43:41 xps9500 kernel:  pm_runtime_work+0x84/0xb0
May 07 19:43:41 xps9500 kernel:  process_one_work+0x1c4/0x380
May 07 19:43:41 xps9500 kernel:  worker_thread+0x51/0x390
May 07 19:43:41 xps9500 kernel:  ? rescuer_thread+0x3b0/0x3b0
May 07 19:43:41 xps9500 kernel:  kthread+0xdb/0x110
May 07 19:43:41 xps9500 kernel:  ? kthread_complete_and_exit+0x20/0x20
May 07 19:43:41 xps9500 kernel:  ret_from_fork+0x1f/0x30
May 07 19:43:41 xps9500 kernel:  </TASK>
#############################
Alt-Sysrq-t
#############################
May 07 19:43:47 xps9500 systemd-journald[367]: Missed 28798 kernel messages
May 07 19:43:47 xps9500 kernel:  S    pool-spawner  2085      2421.445877         7   120         0.000000         0.052103         0.000000         0.000000 0 0 /user.slice/user-1000.slice/user@1000.service/session.slice
May 07 19:43:47 xps9500 kernel:  S    dconf worker  2210      2421.443613        45   120         0.000000         0.580911         0.000000         0.000000 0 0 /user.slice/user-1000.slice/user@1000.service/session.slice
May 07 19:43:47 xps9500 kernel:  S        ibus-x11  2091      2421.574414      2035   120         0.000000       234.479785         0.000000         0.000000 0 0 /user.slice/user-1000.slice/user@1000.service/session.slice
May 07 19:43:47 xps9500 kernel:  S     ibus-portal  2108      2421.443601       151   120         0.000000         7.886734         0.000000         0.000000 0 0 /user.slice/user-1000.slice/user@1000.service/session.slice
May 07 19:43:47 xps9500 kernel:  S       kitty:sh6  2567     20556.057034         8   120         0.000000         0.076481         0.000000         0.000000 0 0 /user.slice/user-1000.slice/user@1000.service/app.slice
May 07 19:43:47 xps9500 kernel:  S xdg-desktop-por  2641      2421.443967        20   120         0.000000         1.410182         0.000000         0.000000 0 0 /user.slice/user-1000.slice/user@1000.service/session.slice
May 07 19:43:47 xps9500 kernel:  S  xdg-de:traceq0  2573     20556.325235         7   139         0.000000         0.092993         0.000000         0.000000 0 0 /user.slice/user-1000.slice/user@1000.service/app.slice


----------
trunced, see https://0x0.st/HZyb.log for full log
----------


#####################################
Running Alt-Sysrq-(r,e,i) which kills most things and restarts DE
#####################################
May 07 19:43:54 xps9500 kernel: sysrq: Keyboard mode set to system default
May 07 19:43:54 xps9500 kernel: sysrq: Terminate All Tasks
May 07 19:43:54 xps9500 systemd-journald[367]: Received SIGTERM.
May 07 19:43:54 xps9500 NetworkManager[819]: <info>  [1683452634.5993] caught SIGTERM, shutting down normally.
May 07 19:43:54 xps9500 wpa_supplicant[951]: p2p-dev-wlp0s20: CTRL-EVENT-DSCP-POLICY clear_all
```

**Comment by YHNdnzj** [comment]
> question, should this be specific to suspend-then-hibernate? or should this be applied to suspend/hibernate as well?

For `suspend-then-hibernate`, we need to periodically wake up the system to check the battery discharging speed or elapsed time in order to determine when to hibernate, while for plain `suspend` or `hibernate` we simply tell the kernel to put us into the corresponding state only once. So yeah, the former requires special handling.

**Comment by YHNdnzj** [comment]
From your kernel backtrace, I think it's not related to freezing/thawing of user slice, but due to a problem in setting up the wakeup timer. Please report this to kernel instead.

**Comment by soraxas** [comment]
Hi there, thank you for your response. I see, yeah that makes sense that it needs to set scheduled wakeups for hibernate.

My apology as I realise the kernel block state *might* be a red herring (I'm not sure why perhaps I was just unlucky on that first try? I should have repeated it...) as I've tried a couple of times after seeing your comment, and it's often without block state (still hangs).

I've recorded its behavior, `suspend-then-hibernate`, within DE terminal and in tty. In tty it actually looks like it's working perfectly.

Video in DE: (when it hangs, I've dump block state and checked journalctl afterwards, and there isn't any block state with `alt-sysrq-w`... strange...)

https://vimeo.com/824773075

Video in tty (seems to works perfectly). I've tried reducing hibernation time to 1 minute, and it does indeed hibernate (requires pressing power key, and booting to grub menu first).

https://vimeo.com/824771721

**Comment by msizanoen1** [comment]
Can you run `systemctl log-level debug` prior to invoking `systemctl suspend-then-hibernate`, and then use `journalctl -b > log.txt` to generate a log file and upload it here?

**Comment by vibertd** [comment]
same issue on fedora 38.

removing the added call to freeze_thaw_user_slice solves the problem.

I wonder whether it is linked somehow to the nvidia driver ?

**Comment by msizanoen1** [comment]
@vibertd Can you run systemctl log-level debug prior to invoking systemctl suspend-then-hibernate, and then use journalctl -b > log.txt to generate a log file and upload it here?

**Comment by mozkomor05** [comment]
Same issue on Arch 6.3.9. Happens only after starting Xorg server (doesn't happen in TTY).

Debug logs: http://ix.io/4znv

**Comment by msizanoen1** [comment]
From the debug logs it seems like the kernel simply failed to suspend at all until you triggered a system wide SIGTERM via sysrq. Can you reproduce this on a non-NVIDIA machine?

**Comment by msizanoen1** [comment]
Also, can you check if the issue is still reproducible after performing the steps at https://wiki.archlinux.org/title/NVIDIA/Tips_and_tricks#Preserve_video_memory_after_suspend?

**Comment by mozkomor05** [comment]
I have already tried Preserving video memory and it didn't help. 

I'll try to run Xorg on integrated Intel graphics. If it doesn't work, I'll try to uninstall Nvidia drivers completely and let you know.

**Comment by msizanoen1** [comment]
> I have already tried Preserving video memory and it didn't help.
> 
> I'll try to run Xorg on integrated Intel graphics. If it doesn't work, I'll try to uninstall Nvidia drivers completely and let you know.

Have you enabled `nvidia-suspend.service` and `nvidia-hibernate.service` as instructed in the Arch wiki? I can't see those units in your debug log.

**Comment by mozkomor05** [comment]
I have disabled them for debug purposes (as I initially thought the suspend-then-hibernate issue was related to them). I've tried enabling them again and I followed the steps listed on Arch wiki with no success. My system didn't even reach the suspend state. I had to SIGTERM which failed the NVIDIA hooks so my laptop didn't suspend at all (which was kind of expected). `systemctl suspend` worked fine even with video memory preservation.

**Comment by mozkomor05** [comment]
I forced Xorg to use intel driver:
```
Section "Device"
        Identifier "intel"
        Driver "intel"
        BusID "PCI:0:2:0"
EndSection
```
Nvidia-smi shows no running processes on GPU:
```
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
```

Suspend-then-hibernate still doesn't work.

**Comment by mozkomor05** [comment]
UPDATE: I fully powered down NVIDIA GPU using udev rules (https://wiki.archlinux.org/title/Hybrid_graphics#Using_udev_rules) and `suspend-then-hibernate` works. You were right about NVIDIA driver failing to suspend.

**Comment by msizanoen1** [comment]
@poettering Would it be OK to temporarily remove the `user.slice` freezing code temporarily until the NVIDIA kernel driver is fixed?

**Comment by bluca** [comment]
> @poettering Would it be OK to temporarily remove the `user.slice` freezing code temporarily upstream until the NVIDIA kernel driver is fixed?

Sorry, no. We don't hack here to fix kernel problems. If your hardware doesn't support suspend/hibernate, then don't use suspend/hibernate until it's fixed.

**Comment by soraxas** [comment]
Ah yep that is correct, I have nvidia non-free drivers installed.

Specificly, I have the prime hybrid intel-nvidia setup so day-to-day I don't really use the nvidia gpu, although it still seems to be affected (perhaps as long as the module is loaded in the kernel?)

**Comment by poettering** [comment]
Anyway, so if this is nvidia closed source related this is unlikely something we'll look into. closed source stuff means you are on your own.

**Comment by soraxas** [comment]
ummm... as mentioned, this is a regression introduced by #24336, so it can hardly be said as not a bug. 

And I think it's hasty to say this is **ONLY** isolated in nvidia's driver (_with your change of title_). At most, you can only say that it is most probable that `freeze_thaw_user_slice(...)` is breaking some mechanisms that Nvidia's driver relies on.

Who's to say that `freeze_thaw_user_slice(...)` is not breaking other, potentially free or non-free, existing drivers?
 (and nvidia driver _just_ happens to be the most common denominator because it happens to be the most common hardwave/drivers out there.)

**Comment by bluca** [comment]
If you know about other drivers that do not work with suspend, then please file a bug upstream to get them fixed

**Comment by soraxas** [comment]
@bluca 

it works with `suspend` but hangs with `suspend-then-hibernate`

**Comment by bluca** [comment]
Sure, same thing: if you are aware of open drivers that do not work, please file a bug with the driver maintainers

**Comment by soraxas** [comment]
How's that the same thing? 

That means that it's not that the driver prevents the system from suspending, but rather the way that the logics with `suspend-then-hibernate` introduced the issue.

**Comment by bluca** [comment]
Do you have a concrete example?

**Comment by soraxas** [comment]
I'm not too sure what you meant by that, other than the regression that had been pointed out earlier.

**Comment by soraxas** [comment]
Um... I mean... it's all just speculation, right? It is equally trivial to point out other common drivers that our systems might also have, and it's none the wise to know what is the main culprit.

The main purpose of my comment was just to point out that the current approach `suspend-then-hibernate` is perhaps breaking some of the assumptions that some existing drivers use (which, we currently **speculates** it's nvidia)

**Comment by poettering** [comment]
nvidia does terrible things in its drivers. We are not going to waste time on making this work, sorry. They are not playing nice with the open source community, and hence we are not interested in investing time to make things work for them.

if they have unpriv userspace hanging around that needs to run between suspend/hibernate cycles, then that's just a very very bad technical choice, but I am certainly not going to invest time on that.

and these things are configurable enough already: you can pick from different choices of suspend/hibernate *already*. I don't think we need to offer more.

**Comment by poettering** [comment]
> So a PR won't be accepted for this?

for what precsiely? that hacks around some nvidia mess? nope. otoh a pr that fixes a bug in systemd that nvidia drivers trigger, but where systemd is at fault and not nivida: yes we'd merge and accept that. but i don't see anything we'd do wrong here…

but systemd is not the place to work around shitty closed source stuff. Sorry.

> I understand that sometimes there might be negative sentiment towards big corp. However, holding these arbitrary moral high grounds can hardly create any tangible damage to those big corps, or have any effect at all. The ones who are left to suffer are often just the end-users.

Oh, i don't care about the moral high ground ore damage to those big corps. I just care about not carrying the burden and having to maintain garbage code. That's all. It's my own self interest as a maintainer.

**Comment by kitwestneat** [comment]
Just to add a bit more context, it appears that nvidia requires that the virtual terminal be changed to an inactive one before suspending. The call to `freeze_thaw_user_slice` causes `chvt` to hang if run in the system-sleep context.

It's possible to work around this by calling chvt before system-sleep is run, but it'd be nice to have everything in a system-sleep script if that's an option.

**Comment by YHNdnzj** [comment]
An environment variable will be added for skipping the freezing/thawing of `user.slice` functionality in #30612.

**Comment by cstrahan** [comment]
In case it helps, this has been my workaround for a while:

https://forums.developer.nvidia.com/t/systemds-suspend-then-hibernate-not-working-in-nvidia-optimus-laptop/213690/4

**Comment by kolayne** [comment]
For those who, just like me, are not yet able to upgrade to systemd 256, there is a workaround: when I switch to a tty and do `systemctl suspend-then-hibernate` from there, things work as expected, with the system properly suspending and resuming. After that, coming back to my graphical tty, it is working fine.

**Comment by kolayne** [comment]
So, I use the following workaround to overcome the suspension/hibernation issues with nvidia drivers installed:

0.  Make sure you have `/usr/bin/chvt` on your system.

1.  Install and enable the service to switch tty before going to sleep:
    ```sh
    $ sudo systemctl edit --full --force tty2_before_sleep.service
    # Paste the following code to the editor and save the file
    [Unit]
    Description=Switch to TTY2 before going to sleep
    Before=sleep.target
    
    [Service]
    Type=oneshot
    User=root
    ExecStart=/usr/bin/chvt 2
    
    [Install]
    WantedBy=sleep.target
    
    $ systemctl enable tty2_before_sleep.service
    ```

2.  Go to sleep normally. Before suspending or hibernating, systemd will run the service and switch you to tty2. It will then suspend/hibernate successfully for my `systemd 256 (256.1-1-arch-g34ba18b^)` and `nvidia 550.90.07-4`.

3.  After resuming, just go back to your graphical tty (usually `Alt+F7` or `Ctrl+Alt+F7` but may be a different tty depending on your graphical environment)

**Comment by kolayne** [comment]
It seems to work for me with `systemd 256 (256.4-1-arch)` and `nvidia 555.58.02-9`
